HIP/hip-xxx.md

```diff
@@ -0,0 +1,191 @@
+
+## Rationale
+
+Hedera's architecture consists of three types of nodes:
+
+- JSON-RPC Relay nodes expose a JSON-RPC API similar to the one that Ethereum nodes use. The Relay handles user requests by either forwarding them to a Consensus or Mirror node.
+- Consensus nodes execute transactions, produce transaction records, and maintain the world state.
+- Mirror nodes ingest the record files produced by the consensus nodes, maintain a “mirror” of the world state, and provide an API for queries and simulations which can work with the current, or a historical state.
+
+Enhancing the JSON-RPC Relay and the Mirror Archive Node with debug APIs will allow users to replay historical transactions and inspect and analyse them by having detailed information for the execution based on a tracer type that is specified in the request. 
+
+Add support of `debug_traceTransaction` in the JSON-RPC Relay which returns detailed information for a historical transaction that was executed in the past.
+
+The JSON-RPC Relay will in turn make a request to the Mirror Node REST API which will be enhanced with a new endpoint that returns opcode logs. The Mirror node already has an endpoint that returns call traces.
+
+## User stories
+
+1. As a user, I want to perform `debug_traceTransaction` calls to explore the op code traces for a historical transaction.
+2. As a user, I want to perform `debug_traceTransaction` calls to explore the sub calls that were executed for a historical transaction.
+
+## Specification
+
+In order to enable trace calls, we should meet some requirements on the Mirror Archive Node side:
+
+- Support EVM versioning, so that we use the appropriate EVM version for a specific historical block.
+- Use full mirror node, keeping state from the genesis block and configured with enabled ingest of `CONTRACT_BYTECODE` , `CONTRACT_STATE_CHANGE` and `CONTRACT_ACTION` sidecar types.
+- Re-execute historical transactions using the information for state values read from the downloaded `CONTRACT_STATE_CHANGE` sidecar record files.
+
+### JSON-RPC Relay Trace API
+
+The JSON-RPC Relay will be enhanced with a `debug_traceTransaction` endpoint. The endpoint has one required parameter - `transactionHash` - and also accepts a list of properties - a `tracer`, and a list of configuration `flags`.
+
+Example payload:
+
+```jsx
+{
+	"jsonrpc": "2.0",
+	"method": "debug_traceTransaction",
+	"params": [
+      // The transaction hash.
+	  "0x123...", 
+	  {
+		// The tracer to use. Accepted values: `callTracer` and `opcodeLogger`. Optional field. Default value is `callTracer`.
+		tracer: "callTracer"
+		// Options for `opcodeLogger`. Ignored for `callTracer`. Disabling some of the metrics collected will reduce the output.
+	    disableMemory: true,
+	    disableStack: true,
+	    disableStorage: true,
+	  },
+	],
+	"id": 0
+}
+```
```

Comment:
Needs an example of the response format we're going for in either case

Reply:
I have added examples for all cases.